File: capstone-projects/cafe-management-program/java/src/Table.java
import java.io.*;
import java.util.*;

public class Table {
    //fields
    private String sID = "TB";
    private String sStatus;
    private List<Order> lOrder;

    public String getsID() {
        return sID;
    }

    public void setsID(String sID) {
        this.sID = sID;
    }

    public String getsStatus() {
        return sStatus;
    }

    public void setsStatus(String sStatus) {
        this.sStatus = sStatus;
    }

    public List<Order> getlOrder() {
        return lOrder;
    }

    //constructor
    public Table() {
        lOrder = new ArrayList<>();
    }

    public Table(String id, String status) {
        sID = id;
        sStatus = status;
        lOrder = new ArrayList<>();
    }

    //Input
    public void InputStatus() {
        System.out.println("Status ([0] Empty - [1] Served): ");
        int st = Main.InputNumber(0, 1);
        switch (st) {
            case 0 -> sStatus = "Empty";
            case 1 -> sStatus = "Served";
        }
    }

    public void Input() {
        if (Cafe.getlTables().size() == 0)
            sID += "0";
        else {
            sID += String.valueOf(FindMaxTableID() + 1);
        }

        System.out.println("Current Table.Table ID: " + sID);
        InputStatus();
    }

    static public int FindMaxTableID() {
        int id = 0;
        for (int i = 0; i < Cafe.getlTables().size(); i++) {
            if (Integer.parseInt(Cafe.getlTables().get(i).getsID().substring(2)) > id)
                id = Integer.parseInt(Cafe.getlTables().get(i).getsID().substring(2));
        }
        return id;
    }

    //Output
    public void Output() {
        System.out.println("\t" + Main.padRight(this.getsID(), 20) + this.getsStatus());
    }

    //Sort
    static public void SortID() {
        for (int i = 0; i < Cafe.getlTables().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlTables().size(); j++) {
                if (Cafe.getlTables().get(i).getsID().compareTo(Cafe.getlTables().get(j).getsID()) > 0) {
                    Table tmp = Cafe.getlTables().get(i);
                    Cafe.getlTables().set(i, Cafe.getlTables().get(j));
                    Cafe.getlTables().set(j, tmp);
                }
            }
        }
    }

    static public void SortStatus() {
        for (int i = 0; i < Cafe.getlTables().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlTables().size(); j++) {
                if (Cafe.getlTables().get(i).getsStatus().compareTo(Cafe.getlTables().get(j).getsStatus()) > 0) {
                    Table tmp = Cafe.getlTables().get(i);
                    Cafe.getlTables().set(i, Cafe.getlTables().get(j));
                    Cafe.getlTables().set(j, tmp);
                }
            }
        }
    }

    //Methods
    //Login_Methods
    static public int SearchTable(String id) {
        id = id.toLowerCase();
        for (int i = 0; i < Cafe.getlTables().size(); i++) {
            if (Cafe.getlTables().get(i).getsID().toLowerCase().equals(id)) {
                return i;
            }
        }
        return -1;
    }

    static public void AddNewTable() {
        System.out.println("\n\t[ADDING NEW TABLE]\n");
        Table tb = new Table();
        tb.Input();
        Cafe.getlTables().add(tb);
    }

    static public int SelectTable() {
        String id;
        int pos;
        do {
            try {
                Scanner sc = new Scanner(System.in);
                System.out.print(" => Now Enter Table ID: ");
                id = sc.nextLine();
                pos = SearchTable(id);
                if (pos == -1)
                    continue;
                else
                    return pos;
            } catch (Exception e) {
                continue;
            }
        } while (true);

    }

    static public void ResetTableID() {
        for (int i = 0; i < Cafe.getlTables().size(); i++) {
            Cafe.getlTables().get(i).setsID("TB" + i);
        }
    }

    //FILES
    static public void WriteDataTable() throws IOException {
        String[] wt = new String[Cafe.getlTables().size()];
        for (int i = 0; i < Cafe.getlTables().size(); i++) {
            wt[i] = Cafe.getlTables().get(i).getsID() + ";" + Cafe.getlTables().get(i).getsStatus();
            if (Cafe.getlTables().get(i).getlOrder().size() != 0) {
                for (int j = 0; j < Cafe.getlTables().get(i).getlOrder().size(); j++) {
                    wt[i] += ";" + Cafe.getlTables().get(i).getlOrder().get(j).getsServiceName()
                            + ";" + Cafe.getlTables().get(i).getlOrder().get(j).getiAmount()
                            + ";" + Cafe.getlTables().get(i).getlOrder().get(j).getdPrice()
                            + ";" + Cafe.getlTables().get(i).getlOrder().get(j).getdCost();
                }
            }
        }
        File file = new File("Table.txt");
        OutputStream outputStream = new FileOutputStream(file);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputStream);
        for (String item : wt) {
            outputStreamWriter.write(item);
            outputStreamWriter.write("\n");
        }
        outputStreamWriter.flush();
    }

    static public void ReadDataTable() throws IOException {

        File file = new File("Table.txt");
        InputStream inputStream = new FileInputStream(file);
        InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
        BufferedReader reader = new BufferedReader(inputStreamReader);

        ArrayList<String> a = new ArrayList<>();
        String line;
        while ((line = reader.readLine()) != null) {
            a.add(line);
        }
        for (String s : a) {
            String[] b = s.split(";");
            Table tb = new Table(b[0], b[1]);
            if (b.length > 2) {
                for (int j = 2; j < b.length; j += 4) {
                    Order od = new Order(b[j],
                            Integer.parseInt(b[j + 1]),
                            Double.parseDouble(b[j + 2]),
                            Double.parseDouble(b[j + 3]));
                    tb.getlOrder().add(od);
                }
            }
            Cafe.getlTables().add(tb);
        }
    }
}

File: capstone-projects/cafe-management-program/cpp/Project_Nhom04_Cpp/Table.cpp
#include "Library.h"

string Table::getsID() {
	return this->sID;
}

void Table::setsID(string sID) {
	this->sID = sID;
}

string Table::getsStatus() {
	return this->sStatus;
}

void Table::setsStatus(string sStatus) {
	this->sStatus = sStatus;
}

Table::Table() {
}

//Input
void Table::InputStatus() {
	printf("Status ([0] Empty - [1] Served): ");
	cout << endl;

	int st = InputNumber(0, 1);
	switch (st) {
	case 0:
		this->sStatus = "Empty";
		break;
	case 1:
		this->sStatus = "Served";
		break;
	}
}

void Table::Input() {
	if (lTables.size() == 0)
		this->sID += "0";
	else {
		this->sID += to_string(FindMaxTableID() + 1);
	}
	cout << "Current Table ID: " << this->sID << endl;
	InputStatus();
}

//Output
void Table::Output() {
	cout << setw(20) << this->getsID() << setw(20) << this->getsStatus() << endl;
}

int FindMaxTableID() {
	int id = 0;
	for (int i = 0; i < lTables.size(); i++) {
		if (stoi(lTables.at(i).getsID().substr(2)) > id)
			id = stoi(lTables.at(i).getsID().substr(2));
	}
	return id;
}

//Sort
void SortTableID() {
	for (int i = 0; i < lTables.size() - 1; i++) {
		for (int j = i + 1; j < lTables.size(); j++) {
			if (lTables.at(i).getsID() > lTables[j].getsID()) {
				Table tmp = lTables.at(i);
				lTables.at(i) = lTables[j];
				lTables[j] = tmp;
			}
		}
	}
}

void SortTableStatus() {
	for (int i = 0; i < lTables.size() - 1; i++) {
		for (int j = i + 1; j < lTables.size(); j++) {
			if (lTables.at(i).getsStatus() > lTables[j].getsStatus()) {
				Table tmp = lTables.at(i);
				lTables.at(i) = lTables[j];
				lTables[j] = tmp;
			}
		}
	}
}

//Methods
//Login_Methods
int SearchTable(string id) {
	id = toLower(id);
	for (int i = 0; i < lTables.size(); i++) {
		if (toLower(lTables.at(i).getsID()) == toLower(id)) {
			return i;
		}
	}
	return -1;
}

void AddNewTable() {
	printf("\n\t[ADDING NEW TABLE]\n");
	Table tb;
	tb.Input();
	lTables.push_back(tb);
}

int SelectTable() {
	string id;
	int pos;
	do {
		try {

			printf(" => Now Enter Table ID: ");
			getline(cin, id);
			pos = SearchTable(id);
			if (pos == -1)
				continue;
			else
				return pos;
		}
		catch (exception e) {
			continue;
		}
	} while (true);

}

void ResetTableID() {
	for (int i = 0; i < lTables.size(); i++) {
		lTables.at(i).setsID("TB" + to_string(i));
	}
}

//FILES
void WriteDataTable() {
	string wt[MAX];
	for (int i = 0; i < lTables.size(); i++) {
		wt[i] = lTables.at(i).getsID() + ";" + lTables.at(i).getsStatus();
		if (lTables.at(i).lOrder.size() > 0) {
			for (int j = 0; j < lTables.at(i).lOrder.size(); j++) {
				wt[i] += ";" + lTables.at(i).lOrder.at(j).getsServiceName()
					+ ";" + to_string(lTables.at(i).lOrder.at(j).getiAmount())
					+ ";" + to_string(lTables.at(i).lOrder.at(j).getdPrice())
					+ ";" + to_string(lTables.at(i).lOrder.at(j).getdCost());
			}
		}
	}
	fstream fileout("Table.txt", std::ios_base::out);
	for (int i = 0; i < lTables.size(); i++) {
		fileout << wt[i];
		if (i < lTables.size() - 1)
			fileout << endl;
	}
	fileout.close();
}

void ReadDataTable() {
	fstream filein("Table.txt", std::ios_base::in);

	while (!filein.eof()) {
		string line[MAX];
		string tmp;
		int dem = 0;
		Table tb;

		getline(filein, tmp);
		if (tmp == "")
		{
			filein.close();
			return;
		}

		dem = SplitString(line, tmp, ';');

		tb.setsID(line[0]);
		tb.setsStatus(line[1]);

		if (dem > 1)
		{
			for (int i = 2; i <= dem; i += 4)
			{
				Order od;
				od.setsServiceName(line[i]);
				od.setiAmount(stoi(line[i + 1]));
				od.setdPrice(stod(line[i + 2]));
				od.setdCost(stod(line[i + 3]));

				tb.lOrder.push_back(od);
			}
		}
		lTables.push_back(tb);
	}
	filein.close();
}